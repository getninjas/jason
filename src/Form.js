import React from 'react';
import { Form, Field } from 'react-final-form';


const onSubmit = () => console.log('ok')
const validate = () => console.log('ok')
const handleSubmit = () => console.log('ok')

const MyForm = ({data}) => {
  return (
    <Form
      render={() => {
        console.log(data);
        return (
          <form onSubmit={onSubmit}>
            <h2>Simple Default Input</h2>
            <div>
              <label>First Name</label>
              <Field name="firstName" component="input" placeholder="First Name" />
            </div>
          </form>
        );
      }}
    />
  );
};

export default MyForm;
