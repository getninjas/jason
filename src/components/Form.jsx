import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Step from './Step';
import Breadcrumb from './Breadcrumb';

export default class Form extends Component {
  constructor() {
    super();

    this.state =  {
      activeStep: 0,
    };
  }

  componentDidMount() {
    this.setState({ activeStep: 0 });
  }

  render() {
    const { action, method, name, form } = this.props;

    return (
      <section className="wall--inverted col-normal-8 col-small-12">
        <form action={action} method={method} name={name} className="form container">
          {
            form.steps.map((step, index) => {
              return <Step visible={this.state.activeStep === index} key={`step-${index}`} step={step} />
            })
          }
        </form>

        <Breadcrumb active={this.state.activeStep} steps={form.steps} />
      </section>
    );
  }
}

Form.defaultProps = {
  form: {
    steps: [],
  }
}

Form.propTypes = {
  action: PropTypes.string,
  method: PropTypes.string,
  name: PropTypes.string.isRequired,
  form: PropTypes.object.isRequired,
}
