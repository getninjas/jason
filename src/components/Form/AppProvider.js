import React, { Component } from 'react';
import PropTypes from 'prop-types';

const AppContext = React.createContext();

class AppProvider extends Component {
  state = {
    greeting : 'hello',
    sayHi: (greeting) => {
      this.setState({greeting: greeting})
    }
  }
  render() {
    return <AppContext.Provider value={this.state}>
    { this.props.children }
    </AppContext.Provider>
  }
}

AppProvider.propTypes = {
  children: PropTypes.any,
}

export default AppProvider;
