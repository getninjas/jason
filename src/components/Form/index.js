import axios from 'axios';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { AppContext } from '../../AppContext';
import Breadcrumb from '../Breadcrumb';
import Step from '../Step';
import { validateField, validateStep } from './validation';

const propTypes = {
  name: PropTypes.string.isRequired,
  data: PropTypes.object.isRequired,
  action: PropTypes.string.isRequired,
  method: PropTypes.string,
  onReady: PropTypes.func,
  handleZipcodeExternalLinkClick: PropTypes.func,
  onZipcodeFetchSuccess: PropTypes.func,
  onZipcodeFetchError: PropTypes.func,
  onSubmitSuccess: PropTypes.func,
  onSubmitFieldError: PropTypes.func,
  onSubmitError: PropTypes.func,
  onSubmit: PropTypes.func,
  onStepChange: PropTypes.func,
};

const defaultProps = {
  method: 'POST',
  onReady() {},
  handleZipcodeExternalLinkClick(event) { return false; },
  onZipcodeFetchSuccess() {},
  onZipcodeFetchError() {},
  onSubmit() {},
  onSubmitSuccess() {},
  onSubmitFieldError() {},
  onSubmitError() {},
  onStepChange() {},
};

export default class Form extends Component {
  constructor(props) {
    super(props);

    this.state = {
      activeStepIndex: 0,
      handleZipcodeExternalLinkClick: (event) => this.handleZipcodeExternalLinkClick(event),
      onZipcodeFetchSuccess: data => this.onZipcodeFetchSuccess(data),
      onZipcodeFetchError: data => this.props.onZipcodeFetchError(data),
      stepsCount: 0,
      address: {
        typeStreet: '',
        street: '',
        neighborhood: '',
        city: '',
        uf: '',
      },
      action: this.props.action,
      steps: [],
    };

    this.formStyle = 'form container sh-form-content space-box-small';
    this.onSubmit = this.onSubmit.bind(this);
    this.onFieldChange = this.onFieldChange.bind(this);
    this.onFieldBlur = this.onFieldBlur.bind(this);
    this.errorMessages = this.props.data.errorMessages;
    this.handleZipcodeExternalLinkClick = this.handleZipcodeExternalLinkClick.bind(this);
  }

  get currentStep() {
    const { steps, activeStepIndex } = this.state;

    return steps[activeStepIndex];
  }

  componentDidMount() {
    this.setState({
      activeStepIndex: 0,
      stepsCount: this.props.data.steps.length - 1,
      steps: this.props.data.steps,
    }, this.props.onReady);
  }

  updateState(state) {
    Object.assign(state, { ...this.state, action: state.action || this.props.action });

    this.setState({ ...state });
  }

  handleZipcodeExternalLinkClick(event) {
    const handleZipcodeExists = this.props && this.props.handleZipcodeExternalLinkClick;
    if (!handleZipcodeExists) { return false; }
    const { handleZipcodeExternalLinkClick } = this.props;
    return handleZipcodeExternalLinkClick(event);
  }

  onZipcodeFetchSuccess(data) {
    this.props.onZipcodeFetchSuccess(data);
    const { typeStreet, street, neighborhood, city, uf } = data;

    this.setState({
      address: {
        typeStreet,
        street,
        neighborhood,
        city,
        uf,
      },
    });
  }

  formSubmit() {
    this.handleStepChange();

    if (this.isLastStep(this.state.activeStepIndex)) {
      this.handleSubmit();
    }
  }

  onSubmit(evt) {
    evt.preventDefault();

    this.formSubmit();
  }

  handleSubmit() {
    if (this.isStepsValid()) {
      this.submitRequest();
    }
  }

  isStepsValid() {
    const errorMessages = this.errorMessages;
    const validSteps = this.state.steps.filter((step) => {
      const { isValid } = validateStep(step.fields, errorMessages);

      return isValid;
    });

    return validSteps.length === this.props.data.steps.length;
  }

  async submitRequest() {
    try {
      this.props.onSubmit();

      const body = this.getFields();
      const response = await axios.post(this.state.action, body);

      this.props.onSubmitSuccess(response);
    } catch (error) {
      this.props.onSubmitError(error);
    }
  }

  getFields() {
    const fields = this.state.steps.map(step => step.fields);

    return { data: { ...fields, address: { ...this.state.address } } };
  }

  beforeStepChange() {
    const currentStep = this.state.steps[this.state.activeStepIndex];

    return currentStep.beforeChange ? currentStep.beforeChange(currentStep) : this.props.onStepChange();
  }

  async handleStepChange() {
    const { updatedFields, isValid } = validateStep(this.currentStep.fields, this.errorMessages);

    this.updateStep(updatedFields);

    if (isValid) {
      await this.beforeStepChange();

      this.nextStep(this.state);

      return;
    }

    const step = this.state.activeStepIndex < this.state.stepsCount ? 'current' : 'last';
    this.props.onSubmitFieldError(step);
  }

  nextStep({ activeStepIndex, stepsCount }) {
    if (activeStepIndex < stepsCount) {
      this.setState({ activeStepIndex: activeStepIndex + 1 });
    }
  }

  onFieldBlur({ value, id, required, type, minLength, regexPattern }) {
    const errorMessages = this.errorMessages;
    const fields = this.currentStep.fields.map((item) => {
      if (item.id === id) {
        const errorMessage = validateField({
          minLength,
          regexPattern,
          required,
          type,
          value,
        }, errorMessages);

        return { ...item, value, errorMessage };
      }

      return item;
    });

    this.updateStep(fields);
  }

  onFieldChange({ value, id }) {
    const fields = this.currentStep.fields.map((item) => {
      if (item.reference === id) {
        const defaultValue = { values: [{ databaseId: '', value: item.mask }] };

        return { ...item, values: (item.nested_values[value] || defaultValue).values, value: null };
      }

      if (item.id === id) {
        return { ...item, value };
      }

      return item;
    });

    this.updateStep(fields);
  }

  updateStep(fields) {
    const { steps, activeStepIndex } = this.state;

    const modifiedSteps = [...steps];
    const modifiedStep = { ...modifiedSteps[activeStepIndex], fields };

    modifiedSteps[activeStepIndex] = modifiedStep;

    this.setState({ steps: modifiedSteps });
  }

  isStepVisible(index) {
    return this.state.activeStepIndex === index;
  }

  isLastStep(index) {
    return index === this.state.stepsCount;
  }

  render() {
    return (
      <AppContext.Provider value={this.state}>
        <form noValidate onSubmit={this.onSubmit} name={this.props.name} action={this.state.action}
          className={this.formStyle}>
          {
            this.state.steps.map((step, index) => {
              const { buttonText, headerMarkup, fields } = step;

              return (
                <Step
                  buttonText={buttonText}
                  fields={fields}
                  formName={this.props.name}
                  onSubmit={this.onSubmit}
                  headerMarkup={headerMarkup}
                  isLast={this.isLastStep(index)}
                  key={`${this.props.name}-step-${index}`}
                  onFieldBlur={this.onFieldBlur}
                  onFieldChange={this.onFieldChange}
                  visible={this.isStepVisible(index)}
                  zipcodeUrlService={this.props.data.zipcodeUrlService}
                />
              );
            })
          }
        </form>

        <Breadcrumb active={this.state.activeStepIndex} steps={this.state.steps} />
      </AppContext.Provider>
    );
  }
}

Form.propTypes = propTypes;
Form.defaultProps = defaultProps;
