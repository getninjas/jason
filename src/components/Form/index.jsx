import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { AppContext } from '../../AppContext';
import Step from '../Step';
import Breadcrumb from '../Breadcrumb';
import { validateField, validateStep } from './validation';

const propTypes = {
  name: PropTypes.string.isRequired,
  data: PropTypes.object.isRequired,
  action: PropTypes.string.isRequired,
  method: PropTypes.string,
};

const defaultProps = {
  method: 'POST',
};

export default class Form extends Component {
  constructor() {
    super();

    this.state = {
      activeStepIndex: 0,
      onZipcodeFetchSuccess: zipcode => window.alert(`zipcode fectch success hook ${zipcode}`),
      onZipcodeFetchError: zipcode => window.alert(`zipcode fectch error hook ${zipcode}`),
      stepsCount: 0,
      steps: [],
    };

    this.sectionStyle = "wall--inverted col-normal-8 col-small-12";
    this.formStyle = "form container sh-form-content space-box-small";

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleButtonClick = this.handleButtonClick.bind(this);
    this.onFieldChange = this.onFieldChange.bind(this);
  }

  get currentStep() {
    const { steps, activeStepIndex } = this.state;

    return steps[activeStepIndex];
  }

  componentDidMount() {
    this.setState({
      activeStepIndex: 0,
      stepsCount: this.props.data.steps.length - 1,
      steps: this.props.data.steps,
    });
  }

  handleSubmit(evt) {
    evt.preventDefault();

    this.handleStepChange();
  }

  handleButtonClick(evt) {
    evt.preventDefault();

    this.handleStepChange();
  }

  handleStepChange() {
    const { updatedFields, isValid } = validateStep(this.currentStep.fields);

    this.updateStep(updatedFields);

    if (isValid) {
      this.nextStep(this.state);
    }
  }

  nextStep({ activeStepIndex, stepsCount }) {
    if (activeStepIndex < stepsCount) {
      this.setState({ activeStepIndex: activeStepIndex + 1 });
    }
  }

  onFieldChange({ value, id, required, type, minLength }) {
    const fields = this.currentStep.fields.map((item) => {
      const itemID = `${this.props.name}-${item.id}`;

      if (itemID === id) {
        const errorMessage = validateField({ required, type, value, minLength });

        return Object.assign({}, item, { value, errorMessage });
      }

      return item;
    });

    this.updateStep(fields);
  }

  updateStep(fields) {
    const { steps, activeStepIndex } = this.state;

    let modifiedSteps = [...steps];
    let modifiedStep = Object.assign({}, modifiedSteps[activeStepIndex], { fields });

    modifiedSteps[activeStepIndex] = modifiedStep;

    this.setState({
      steps: modifiedSteps,
    });
  }

  isStepVisible(index) {
    return this.state.activeStepIndex === index;
  }

  isLastStep(index) {
    return index === this.state.stepsCount;
  }

  render() {
    const { action, method, name } = this.props;

    return (
      <AppContext.Provider value={this.state}>
        <section className={this.sectionStyle}>
          <form noValidate onSubmit={this.handleSubmit} action={action} method={method} name={name} className={this.formStyle}>
            {
              this.state.steps.map((step, index) => {
                const { buttonText, headerMarkup, fields } = step;

                return (
                  <Step
                    buttonText={buttonText}
                    fields={fields}
                    formName={name}
                    handleButtonClick={this.handleButtonClick}
                    headerMarkup={headerMarkup}
                    isLast={this.isLastStep(index)}
                    key={`${name}-step-${index}`}
                    onFieldChange={this.onFieldChange}
                    visible={this.isStepVisible(index)}
                    zipcodeUrlService={this.props.data.zipcodeUrlService}
                  />
                )
              })
            }
          </form>

          <Breadcrumb active={this.state.activeStepIndex} steps={this.state.steps} />
        </section>
      </AppContext.Provider>
    );
  }
}

Form.propTypes = propTypes;
Form.defaultProps = defaultProps;
