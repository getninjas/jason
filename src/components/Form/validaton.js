import validator from 'validator';

export const validateValue = ({ required, type, value }) => {
  if (required && value === null) {
    return 'This field is required';
  }

  if (type === 'phone' && validator.isEmpty(value)) {
    return 'Invalid phone';
  }

  if (type === 'email' && !validator.isEmail(value)) {
    return 'Invalid email';
  }

  if (required && validator.isEmpty(value)) {
    return 'This field is required';
  }

  return '';
}

export const validateStep = (fields)=> {
  let isValid = true;

  const updatedFields = fields.map((field) => {
    const modifiedField = Object.assign({}, field);
    const errorMessage = validateValue(modifiedField);

    if (errorMessage) {
      modifiedField.errorMessage = errorMessage;
      isValid = false;
    }

    return modifiedField;
  });

  return { updatedFields, isValid };
}
