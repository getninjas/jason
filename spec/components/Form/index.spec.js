import React from 'react';
import renderer from 'react-test-renderer';
import Form from '../../../src/components/Form';
import { form } from '../../../src/form.json';
import { enzymeConfig, shallow } from '../../enzymeConfig';
import fillFormFields, { sleep } from '../../helper';

enzymeConfig();

function createNodeMock(element) {
  if (element.type === 'input') {
    return {
      addEventListener() { },
      value: '',
    };
  }

  return null;
}

const copyState = data => (JSON.parse(JSON.stringify(data)));

describe('Form', () => {
  it('renders defaultProps ', () => {
    const data = copyState(form);
    const options = { createNodeMock };
    const component = renderer.create(
      <Form name={'form'} action={'/'} data={data} />, options,
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('.formSubmit', () => {
    describe('when is not the last step', () => {
      it('does not call .handleSubmit', () => {
        const data = copyState(form);
        const component = shallow(
          <Form name={'form'} action={'/'} data={data} />,
        );
        const instance = component.instance();

        instance.handleStepChange = jest.fn();
        instance.isLastStep = jest.fn();
        instance.handleSubmit = jest.fn();

        instance.formSubmit();

        expect(instance.handleStepChange).toBeCalled();
        expect(instance.isLastStep).toBeCalledWith(0);
        expect(instance.handleSubmit).not.toBeCalled();
      });
    });

    describe('when is last step', () => {
      it('calls .handleSubmit', () => {
        const data = copyState(form);
        data.steps = fillFormFields(data.steps);
        data.steps[1].fields[2].required = false;

        const component = shallow(
          <Form name={'form'} action={'/'} data={data} />,
        );
        const instance = component.instance();
        const evt = { preventDefault() { } };

        instance.handleSubmit = jest.fn();

        data.steps.map(() => instance.nextStep(component.state()));

        instance.onSubmit(evt);

        expect(instance.handleSubmit).toBeCalled();
      });
    });
  });

  describe('.getFields', () => {
    it('returns all form fields plus address object', () => {
      const data = copyState(form);
      const address = {
        type_street: 'type_street',
        street: 'street',
        neighborhood: 'neighborhood',
        city: 'city',
        uf: 'uf',
      };

      data.steps = fillFormFields(data.steps);

      const component = shallow(
        <Form name={'form'} action={'/'} data={data} />,
      );

      const fields = component.state().steps.map(step => step.fields);
      const mockFields = { data: { ...fields, address: { ...address } } };

      component.state().address = { ...address };

      const result = component.instance().getFields();

      expect(result).toEqual(mockFields);
    });
  });

  describe('.updateState', () => {
    describe('with prefilled zipcode', () => {
      it('matches updated state', () => {
        const data = copyState(form);
        const component = shallow(
          <Form name={'form'} data={data} action={'/'} />,
        );

        const mock = [{
          title: 'Zipcode',
          required: false,
          type: 'zipcode',
          id: 'zipcode',
          name: 'zipcode',
          placeholder: '00000-000',
          value: '04904-160',
          wrapperClassName: 'form__field input',
          values: [],
        }];

        const last = data.steps.length - 1;
        const userFields = data.steps[last];

        Object.assign(userFields, { fields: [...mock] });

        component.instance().updateState(data);

        expect(component.state().steps[last].fields).toBe(userFields.fields);
      });
    });

    describe('with default values', () => {
      it('does not change state', () => {
        const data = copyState(form);
        const component = shallow(
          <Form name={'form'} data={data} action={'/'} />,
        );

        component.instance().updateState(data);

        expect(component.state()).toEqual(data);
      });
    });
  });

  describe('.submitRequest', () => {
    it('calls .props.onSubmitSuccess, getFields', async () => {
      const onSubmit = jest.fn();
      const onSubmitSuccess = jest.fn();
      const component = shallow(
        <Form name={'form'} onSubmit={onSubmit} onSubmitSuccess={onSubmitSuccess} action={'/'} data={form} />,
      );

      component.instance().getFields = jest.fn();

      await component.instance().submitRequest();

      expect(component.instance().props.onSubmit).toBeCalled();
      expect(component.instance().getFields).toBeCalled();
      expect(component.instance().props.onSubmitSuccess).toBeCalled();
    });
  });

  describe('.isStepVisible', () => {
    const component = shallow(
      <Form name={'form'} action={'/'} data={form} />,
    );

    it('returns true for activeStepIndex', () => {
      const stepOneIsVisible = component
        .instance()
        .isStepVisible(component.state().activeStepIndex);

      expect(stepOneIsVisible).toBe(true);
    });

    it('returns false for secondStep', () => {
      const stepTwoIsVisible = component
        .instance()
        .isStepVisible(component.state().activeStepIndex + 1);

      expect(stepTwoIsVisible).toBe(false);
    });
  });

  describe('.isLastStep', () => {
    const component = shallow(
      <Form name={'form'} action={'/'} data={form} />,
    );

    it('returns false for firstStep ', () => {
      const stepOne = component.state().activeStepIndex;
      const stepOneIsLast = component.instance().isLastStep(stepOne);

      expect(stepOneIsLast).toBe(false);
    });

    it('returns true for thirdStep ', () => {
      const stepThree = component.state().activeStepIndex + 2;
      const stepThreeIsLast = component.instance().isLastStep(stepThree);

      expect(stepThreeIsLast).toBe(true);
    });
  });

  describe('.nextStep', () => {
    const component = shallow(
      <Form name={'form'} action={'/'} data={form} />,
    );

    it('moves to nextStep', () => {
      const currentStepIndex = component.state().activeStepIndex;

      component.instance().nextStep(component.state());

      expect(component.state().activeStepIndex).toBe(currentStepIndex + 1);
    });

    it('stays on the currentStep', () => {
      component.instance().nextStep(component.state());

      const currentStepIndex = component.state().activeStepIndex;
      component.instance().nextStep(component.state());

      expect(currentStepIndex).toBe(component.state().activeStepIndex);
    });
  });

  describe('.handleStepChange', () => {
    it('calls .updateStep', () => {
      const onSubmitFieldError = jest.fn();
      const component = shallow(
        <Form name={'form'} action={'/'} data={form} onSubmitFieldError={onSubmitFieldError} />,
      );

      component.instance().updateStep = jest.fn();
      component.instance().nextStep = jest.fn();

      component.instance().handleStepChange();

      expect(component.instance().updateStep).toBeCalled();
      expect(component.instance().nextStep).not.toBeCalled();
      expect(component.instance().props.onSubmitFieldError).toBeCalled();
    });

    it('calls .nextStep', async () => {
      const data = copyState(form);
      const onSubmitFieldError = jest.fn();

      data.steps = fillFormFields(data.steps);

      const component = shallow(
        <Form name={'form'} action={'/'} data={data} onSubmitFieldError={onSubmitFieldError} />,
      );

      component.instance().updateStep = jest.fn();
      component.instance().nextStep = jest.fn();

      component.instance().handleStepChange();

      await sleep(100);

      expect(component.instance().updateStep).toBeCalled();
      expect(component.instance().nextStep).toBeCalled();
      expect(component.instance().props.onSubmitFieldError).not.toBeCalled();
    });
  });

  describe('.onFieldChange', () => {
    it('calls .updateStep', () => {
      const component = shallow(
        <Form name={'form'} action={'/'} data={form} />,
      );

      component.instance().updateStep = jest.fn();

      const evt = { preventDefault() { } };
      component.instance().onSubmit(evt);

      expect(component.instance().updateStep).toBeCalled();
    });

    it('expects initial state to differ from current state', async () => {
      const data = copyState(form);

      data.steps = fillFormFields(data.steps);

      const component = shallow(
        <Form name={'form'} action={'/'} data={data} />,
      );

      const evt = { preventDefault() { } };
      const instance = component.instance();

      instance.onSubmit(evt);

      await sleep(100);

      const initialSecondStepState = instance.state;
      const field = { value: '', id: '2_id' };
      instance.onFieldChange(field);

      const updatedState = instance.state;

      expect(updatedState).not.toEqual(initialSecondStepState);
    });

    it('expects updated select with dynamic options', async () => {
      const data = copyState(form);

      data.steps = fillFormFields(data.steps);

      const component = shallow(
        <Form name={'form'} action={'/'} data={data} />,
      );
      const evt = { preventDefault() { } };

      component.instance().onSubmit(evt);

      await sleep(100);

      const updatedOptions = form.steps[1].fields[2].nested_values['124'].values;
      const field = { value: 124, id: '10_id' };

      component.instance().onFieldChange(field);

      const updatedState = component.instance().state;

      expect(updatedState.steps[1].fields[2].values).toEqual(updatedOptions);
    });

    it('does not updated select with dynamic options', () => {
      const component = shallow(
        <Form name={'form'} action={'/'} data={form} />,
      );
      const initialSecondStepState = form.steps[1].fields[2].values;
      const field = { value: '', id: '10_id' };

      component.instance().onFieldChange(field);

      const updatedState = component.instance().state;

      expect(updatedState.steps[1].fields[2].values).toEqual(initialSecondStepState);
    });
  });

  describe('.handleSubmit', () => {
    it('calls .isStepsValid', () => {
      const component = shallow(
        <Form name={'form'} action={'/'} data={form} />,
      );
      component.instance().isStepsValid = jest.fn();

      const evt = { preventDefault() { } };
      component.instance().handleSubmit(evt);

      expect(component.instance().isStepsValid).toBeCalled();
    });

    it('calls .submitRequest when step is valid', () => {
      const CONTACT_FIELD = 2;
      const data = copyState(form);
      data.steps = fillFormFields(data.steps);
      data.steps[CONTACT_FIELD].fields[0].value = '11111-111';
      data.steps[CONTACT_FIELD].fields[2].value = 'iondr@ig.com';
      data.steps[CONTACT_FIELD].fields[3].value = '(11) 98888-9999';

      const component = shallow(
        <Form name={'form'} action={'/'} data={data} />,
      );

      component.instance().submitRequest = jest.fn();
      component.instance().handleSubmit();

      expect(component.instance().submitRequest).toHaveBeenCalled();
    });

    it('does not display next step', () => {
      const component = shallow(
        <Form name={'form'} action={'/'} data={form} />,
      );
      const evt = { preventDefault() { } };
      component.instance().handleSubmit(evt);

      const { activeStepIndex } = component.state();

      expect(activeStepIndex).toEqual(0);
    });

    it('goes to next step', async () => {
      const data = copyState(form);
      data.steps = fillFormFields(data.steps);

      const component = shallow(
        <Form name={'form'} action={'/'} data={data} />,
      );

      const initialStep = component.state().activeStepIndex;

      component.instance().formSubmit();

      await sleep(100);
      const { activeStepIndex } = component.state();

      expect(initialStep).toEqual(0);
      expect(activeStepIndex).toEqual(initialStep + 1);
    });
  });
});
