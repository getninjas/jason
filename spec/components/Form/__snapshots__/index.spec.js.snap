// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders defaultProps 1`] = `
<section
  className="wall--inverted col-normal-8 col-small-12"
>
  <form
    action="/"
    className="form container sh-form-content space-box-small"
    data={
      Object {
        "databaseId": 81,
        "steps": Array [
          Object {
            "breadcrumbTitle": "Your request",
            "buttonText": "Start request",
            "fields": Array [
              Object {
                "id": "1_id",
                "name": "1_name_select",
                "placeholder": "Placeholder",
                "required": true,
                "title": "Qual é o tipo de serviço que você precisa?",
                "type": "select",
                "value": null,
                "values": Array [
                  Object {
                    "databaseId": 7117,
                    "value": "Construção",
                  },
                  Object {
                    "databaseId": 7118,
                    "value": "Reformas",
                  },
                  Object {
                    "databaseId": 7119,
                    "value": "Instalações",
                  },
                  Object {
                    "databaseId": 7120,
                    "value": "Troca",
                  },
                  Object {
                    "databaseId": 7121,
                    "value": "Outros",
                  },
                ],
              },
              Object {
                "id": "2_id",
                "name": "2_name_text",
                "placeholder": "Bora pra action",
                "required": true,
                "title": "Especifique melhor a metragem do serviço",
                "type": "text",
                "value": null,
                "values": Array [],
              },
              Object {
                "id": "3_id",
                "name": "3_name_textarea",
                "placeholder": "Descreva o que você precisa",
                "required": true,
                "title": "Informações adicionais",
                "type": "textarea",
                "value": null,
                "values": Array [],
              },
            ],
            "headerMarkup": "<h2 class='widget__title section-title--high-contrast'>Lorem Ipsum is simply dummy text <small>It has survived not</small></h2>",
            "title": "General data",
          },
          Object {
            "breadcrumbTitle": "Your data",
            "buttonText": "Finish",
            "fields": Array [
              Object {
                "id": "4_id",
                "name": "4_zipcode",
                "placeholder": "00000-000",
                "required": true,
                "title": "Zipcode",
                "type": "zipcode",
                "value": null,
                "values": Array [],
              },
              Object {
                "id": "user_field_2",
                "name": "user_field_2",
                "placeholder": "",
                "required": true,
                "title": "Nome",
                "type": "text",
                "value": null,
                "values": Array [],
              },
              Object {
                "id": "user_field_3",
                "name": "user_field_3",
                "placeholder": "Ex: nome@email.com",
                "required": true,
                "title": "E-mail",
                "type": "email",
                "value": null,
                "values": Array [],
              },
              Object {
                "id": "user_field_4",
                "name": "user_field_4",
                "placeholder": "",
                "required": true,
                "title": "DDD + Celular",
                "type": "phone",
                "value": null,
                "values": Array [],
              },
            ],
            "headerMarkup": "<div class='form__alert-message space-element-large'><img src='https://www.getninjas.com.br/assets/request/phone_call.svg' class='form__icon' alt='Lorem Ipsum is simply dummy text of the printing and typesetting industry.'><h2 class='form__info'>Almost there<small>Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's<strong>standard dummy text ever since the 1500s</strong></small></h2></div>",
            "title": "User Data",
          },
        ],
        "title": "Form Title",
        "zipcodeUrlService": "https://www.dummy.com/zipcode?q=@@zipcode@@",
      }
    }
    method="POST"
    name="form"
    noValidate={true}
    onSubmit={[Function]}
  >
    <fieldset
      className="form__container inputs"
      style={
        Object {
          "display": "block",
        }
      }
    >
      <div
        className="__headerMarkup__"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<h2 class='widget__title section-title--high-contrast'>Lorem Ipsum is simply dummy text <small>It has survived not</small></h2>",
          }
        }
      />
      <div
        className="form__field form__field--fluid input"
      >
        <label
          className="form__label"
          htmlFor="1_id"
        >
          Qual é o tipo de serviço que você precisa?
        </label>
        <select
          className="form__input"
          defaultValue={null}
          id="form-1_id"
          name="1_name_select"
          onChange={[Function]}
          required="true"
        >
          <option
            value=""
          >
            Placeholder
          </option>
          <option
            value={7117}
          >
            Construção
          </option>
          <option
            value={7118}
          >
            Reformas
          </option>
          <option
            value={7119}
          >
            Instalações
          </option>
          <option
            value={7120}
          >
            Troca
          </option>
          <option
            value={7121}
          >
            Outros
          </option>
        </select>
        <span
          className="form__message form__message--invalid space-element-small error"
        >
          
        </span>
      </div>
      <div
        className="form__field form__field--fluid input"
      >
        <label
          className="form__label"
          htmlFor="2_id"
        >
          Especifique melhor a metragem do serviço
        </label>
        <input
          className="form__input"
          id="form-2_id"
          maxLength={255}
          minLength={3}
          name="2_name_text"
          onChange={[Function]}
          placeholder="Bora pra action"
          required="true"
          title=""
          type="text"
          value=""
        />
        <span
          className="form__message form__message--invalid space-element-small error"
        >
          
        </span>
      </div>
      <div
        className="form__field form__field--fluid input"
      >
        <label
          className="form__label"
          htmlFor="3_id"
        >
          Informações adicionais
        </label>
        <textarea
          className="form__input"
          id="form-3_id"
          maxLength={255}
          minLength={3}
          name="3_name_textarea"
          onChange={[Function]}
          placeholder="Descreva o que você precisa"
          required={true}
          title=""
          value=""
        />
        <span
          className="form__message form__message--invalid space-element-small error"
        >
          
        </span>
      </div>
      <button
        className="btn btn--regular btn--high-contrast btn--fluid space-box-medium"
        onClick={[Function]}
        type="button"
      >
        Start request
      </button>
    </fieldset>
    <fieldset
      className="form__container inputs"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="__headerMarkup__"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<div class='form__alert-message space-element-large'><img src='https://www.getninjas.com.br/assets/request/phone_call.svg' class='form__icon' alt='Lorem Ipsum is simply dummy text of the printing and typesetting industry.'><h2 class='form__info'>Almost there<small>Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's<strong>standard dummy text ever since the 1500s</strong></small></h2></div>",
          }
        }
      />
      <div
        className="form__field form__field--fluid input"
      >
        <label
          className="form__label"
          htmlFor="4_id"
        >
          Zipcode
        </label>
        <a
          className="form__label-link"
          href="http://www.buscacep.correios.com.br"
          rel="noopener noreferrer"
          target="_blank"
        >
          Não lembra seu CEP?
        </a>
        <input
          className="form__input"
          id="form-4_id"
          name="4_zipcode"
          onChange={[Function]}
          placeholder="00000-000"
          required={true}
          type="tel"
        />
        <span
          className="full-address"
        >
          
        </span>
        <input
          id="street"
          name="street"
          type="hidden"
          value=""
        />
        <input
          id="neighborhood"
          name="neighborhood"
          type="hidden"
          value=""
        />
        <input
          id="city"
          name="city"
          type="hidden"
          value=""
        />
        <input
          id="uf"
          name="uf"
          type="hidden"
          value=""
        />
        <span
          className="form__message form__message--invalid space-element-small error"
        >
          
        </span>
      </div>
      <div
        className="form__field form__field--fluid input"
      >
        <label
          className="form__label"
          htmlFor="user_field_2"
        >
          Nome
        </label>
        <input
          className="form__input"
          id="form-user_field_2"
          maxLength={255}
          minLength={3}
          name="user_field_2"
          onChange={[Function]}
          placeholder=""
          required="true"
          title=""
          type="text"
          value=""
        />
        <span
          className="form__message form__message--invalid space-element-small error"
        >
          
        </span>
      </div>
      <div
        className="form__field form__field--fluid input"
      >
        <label
          className="form__label"
          htmlFor="user_field_3"
        >
          E-mail
        </label>
        <input
          className="form__input"
          id="form-user_field_3"
          maxLength={255}
          minLength={3}
          name="user_field_3"
          onChange={[Function]}
          placeholder="Ex: nome@email.com"
          required="true"
          title=""
          type="email"
          value=""
        />
        <span
          className="form__message form__message--invalid space-element-small error"
        >
          
        </span>
      </div>
      <div
        className="form__field form__field--fluid input"
      >
        <label
          className="form__label"
          htmlFor="user_field_4"
        >
          DDD + Celular
        </label>
        <input
          className="form__input"
          id="form-user_field_4"
          maxLength={255}
          minLength={3}
          name="user_field_4"
          onChange={[Function]}
          placeholder=""
          required="true"
          title=""
          type="tel"
          value=""
        />
        <span
          className="form__message form__message--invalid space-element-small error"
        >
          
        </span>
      </div>
      <button
        className="btn btn--regular btn--high-contrast btn--fluid space-box-medium"
        onClick={[Function]}
        type="submit"
      >
        Finish
      </button>
    </fieldset>
  </form>
  <ol
    className="form__steps space-box-medium"
  >
    <li
      className="form__steps-item form__steps-item--active"
    >
      Your request
    </li>
    <li
      className="form__steps-item "
    >
      Your data
    </li>
  </ol>
</section>
`;
